File: PixelSolution/PixelEngine/Rigidbody2D.cpp
#include "pch.h"
#include "Rigidbody2D.h"
#include "Transform.h"
#include "PixelEngineAPI.h"
#include "Registry.h"
#include "PixelEngine.h"
#include "PhysManager.h"
#include "ColliderFactory.h"
#include "Physics2DData.h"
#include "PixelMetaAPI.h"

extern PixelEngine* Engine;
Rigidbody2DData* Rigidbody2D_Add(unsigned int id)
{
	auto registry = GetRegistry();
	registry->Add<Rigidbody2DData>(id);
	registry->Add<Physics2DData>(id);
	auto data1 = registry->Get<Rigidbody2DData>(id);
	auto data2 = registry->Get<Physics2DData>(id);
	data1->thisID = id;
	data2->thisID = id;
	return data1;
}

Rigidbody2DData* Rigidbody2D_Get(unsigned int id)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Rigidbody2DData>(id);
	return data;
}

bool Rigidbody2D_Has(unsigned int id)
{
	auto registry = GetRegistry();
	return registry->Has<Rigidbody2DData>(id);
}

void Rigidbody2D_CreateBoxCollider(unsigned int id, PVector2 size, PVector2 center)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	data1->colliderRefC = ColliderFactory::CreateBox2D(&size, &center);
	data1->colliderType = Physics2DData::BOX;
}

void Rigidbody2D_CreateCircleCollider(unsigned int id, float radius, PVector2 center)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	data1->colliderRefC = ColliderFactory::CreateCircle2D(radius, &center);
	data1->colliderType = Physics2DData::CIRCLE;
}

void Rigidbody2D_SetMotionType(unsigned int id, MotionType motionType)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);

	switch (motionType)
	{
	case MotionType::Kinematic:
		data1->Motiontype = JPH::EMotionType::Kinematic;
		break;
	case MotionType::Dynamic:
		data1->Motiontype = JPH::EMotionType::Dynamic;
		break;
	case MotionType::Static:
		data1->Motiontype = JPH::EMotionType::Static;
		break;
	}

	if (data1->bodyID.IsInvalid() == true)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetMotionType(data1->bodyID, data1->Motiontype);
	}
}

void Rigidbody2D_SetPosition(unsigned int id, PVector3 position)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetPosition(data1->bodyID, position.x, position.y, position.z);
	}
	auto transform = registry->Get<TransformData>(id);
	transform->position.x = position.x;
	transform->position.y = position.y;
	transform->position.z = position.z;
}

void Rigidbody2D_SetRotation(unsigned int id, PVector3 rotation)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetRotation(data1->bodyID, rotation.x, rotation.y, rotation.z);
	}
	auto transform = registry->Get<TransformData>(id);
	transform->rotation.x = rotation.x;
	transform->rotation.y = rotation.y;
	transform->rotation.z = rotation.z;
}

void Rigidbody2D_SetVelocity(unsigned int id, PVector3 velocity)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetVelocity(data1->bodyID, velocity.x, velocity.y, velocity.z);
	}
	else
	{
		if (data1->Create == true)
		{
			PixelLog::Error("[Rigidbody2D][SetVelocity] Not Find Collider");
		}
	}
}

void Rigidbody2D_SetGravity(unsigned int id, float gravity)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetGravity(data1->bodyID, gravity);
	}
	else
	{
		data1->gravity = gravity;
	}
}

void Rigidbody2D_SetRestitution(unsigned int id, float restitution)
{

}

void Rigidbody2D_SetFriction(unsigned int id, float restitution)
{

}

void Rigidbody2D_SetLinearDamping(unsigned int id, float linearDamping)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);

	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetLinearDamping(data1->bodyID, linearDamping);
	}
	else
	{
		data1->linearDamping = linearDamping;
	}
}

void Rigidbody2D_SetSensor(unsigned int id, bool sensor)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetSensor(data1->bodyID, sensor);
	}
	data1->sensor = sensor;
}

void Rigidbody2D_SetActive(unsigned int id, bool active)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetActive(data1->bodyID, active);
	}
	data1->active = active;
}

void Rigidbody2D_SetLayer(unsigned int id, const char* layerName)
{
	auto registry = GetRegistry();
	auto data1 = registry->Get<Physics2DData>(id);
	if (data1 == nullptr || layerName == nullptr)
	{
		PixelLog::Error("[Rigidbody2D][SetLayer] Invalid Component Or LayerName");
		return;
	}
	if (data1->bodyID.IsInvalid() == false)
	{
		auto phys = Engine->GetFactory<PhysicsFunction>();
		phys->SetLayer(data1->bodyID, layerName);
	}
	data1->layer = layerName;
}

const char* Rigidbody2D_GetLayer(unsigned int id)
{
	auto data = GetRegistry()->Get<Physics2DData>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[Rigidbody2D][GetLayer] Not Find Component");
		return "";
	}
	return data->layer.c_str();
}

File: Survival/Engine/Rigidbody2D.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

ffi.cdef[[
	typedef struct {
		unsigned int thisID;
	} Rigidbody2DData;

	Rigidbody2DData* Rigidbody2D_Add(unsigned int ID);
	Rigidbody2DData* Rigidbody2D_Get(unsigned int ID);
	bool Rigidbody2D_Has(unsigned int ID);
	void Rigidbody2D_CreateBoxCollider(unsigned int ID,PVector2 Size,PVector2 Center);
	void Rigidbody2D_CreateCircleCollider(unsigned int ID,float Radius,PVector2 Center);
	void Rigidbody2D_SetMotionType(unsigned int ID,MotionType MotionType);
	void Rigidbody2D_SetPosition(unsigned int ID,PVector3 position);
	void Rigidbody2D_SetRotation(unsigned int ID,PVector3 rotation);
	void Rigidbody2D_SetVelocity(unsigned int ID,PVector3 velocity);
	void Rigidbody2D_SetGravity(unsigned int ID,float Gravity);
	void Rigidbody2D_SetRestitution(unsigned int ID,float Restitution);
	void Rigidbody2D_SetFriction(unsigned int ID,float Friction);
	void Rigidbody2D_SetLinearDamping(unsigned int ID,float LinearDamping);
	void Rigidbody2D_SetSensor(unsigned int ID,bool Sensor);
	void Rigidbody2D_SetActive(unsigned int ID,bool Active);
	void Rigidbody2D_SetLayer(unsigned int ID,const char* LayerName);
	const char* Rigidbody2D_GetLayer(unsigned int ID);
]]

---@class Rigidbody2DData
local Rigidbody2DData = {}

---@param Size PVector2
---@param Center PVector2
function Rigidbody2DData:CreateBoxCollider(Size,Center)
	return dll.Rigidbody2D_CreateBoxCollider(self.thisID,Size,Center)
end

---@param Radius number
---@param Center PVector2
function Rigidbody2DData:CreateCircleCollider(Radius,Center)
	return dll.Rigidbody2D_CreateCircleCollider(self.thisID,Radius,Center)
end

---@param MotionType MotionType
function Rigidbody2DData:SetMotionType(MotionType)
	return dll.Rigidbody2D_SetMotionType(self.thisID,MotionType)
end

---@param position PVector3
function Rigidbody2DData:SetPosition(position)
	return dll.Rigidbody2D_SetPosition(self.thisID,position)
end

---@param rotation PVector3
function Rigidbody2DData:SetRotation(rotation)
	return dll.Rigidbody2D_SetRotation(self.thisID,rotation)
end

---@param velocity PVector3
function Rigidbody2DData:SetVelocity(velocity)
	return dll.Rigidbody2D_SetVelocity(self.thisID,velocity)
end

---@param Gravity number
function Rigidbody2DData:SetGravity(Gravity)
	return dll.Rigidbody2D_SetGravity(self.thisID,Gravity)
end

---@param Restitution number
function Rigidbody2DData:SetRestitution(Restitution)
	return dll.Rigidbody2D_SetRestitution(self.thisID,Restitution)
end

---@param Friction number
function Rigidbody2DData:SetFriction(Friction)
	return dll.Rigidbody2D_SetFriction(self.thisID,Friction)
end

---@param LinearDamping number
function Rigidbody2DData:SetLinearDamping(LinearDamping)
	return dll.Rigidbody2D_SetLinearDamping(self.thisID,LinearDamping)
end

---@param Sensor boolean
function Rigidbody2DData:SetSensor(Sensor)
	return dll.Rigidbody2D_SetSensor(self.thisID,Sensor)
end

---@param Active boolean
function Rigidbody2DData:SetActive(Active)
	return dll.Rigidbody2D_SetActive(self.thisID,Active)
end

---@param LayerName string
function Rigidbody2DData:SetLayer(LayerName)
	return dll.Rigidbody2D_SetLayer(self.thisID,LayerName)
end

---@return string
function Rigidbody2DData:GetLayer()
	return ffi.string(dll.Rigidbody2D_GetLayer(self.thisID))
end

local Rigidbody2D_mt = {
	__index = Rigidbody2DData
}

ffi.metatype("Rigidbody2DData", Rigidbody2D_mt)

Rigidbody2D = Rigidbody2D or {}

---@param ID number 
---@return Rigidbody2DData 
function Rigidbody2D.Add(ID) 
	local data = dll.Rigidbody2D_Add(ID)
	data.thisID = ID
	return data 
end 

---@param ID number 
---@return Rigidbody2DData 
function Rigidbody2D.Get(ID) 
	return dll.Rigidbody2D_Get(ID) 
end 

---@param ID number 
---@return boolean 
function Rigidbody2D.Has(ID) 
	return dll.Rigidbody2D_Has(ID) 
end 

return Rigidbody2D